File: 206_reverse_linked_list.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
    def __str__(self):
        cur = self
        res = "Linked List: "
        while cur:
            res += str(cur.val) + " "
            cur = cur.next
        return res

class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        res = None
        while head:
            temp = ListNode(head.val, res)
            res = temp
            head = head.next

        return res

File: c++/206_reverse_linked_list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode* cur_node = head;
        ListNode* prev = nullptr;

        while (cur_node != nullptr) {
            ListNode* temp = cur_node->next;
            cur_node->next = prev;
            prev = cur_node;
            cur_node = temp;
        }
        
        return prev;
    }
};
